File: proga/1_semestr/labs/lab3/src/main/java/dev/lysmux/lab3/entity/item/MoonStone.java
package dev.lysmux.lab3.entity.item;

import dev.lysmux.lab3.common.Case;
import dev.lysmux.lab3.entity.ability.GlowsInDark;

import java.util.Objects;

public class MoonStone extends Item {
    @Override
    public String getPluralName() {
        return "лунные камни";
    }

    @Override
    public String caseDeclension(Case to_case) {
        return switch (to_case) {
            case NOMINATIVE, ACCUSATIVE -> "лунный камень";
            case GENITIVE -> "лунного камня";
            case DATIVE -> "лунному камню";
            case CREATIVE -> "лунным камнем";
            case PREPOSITIONAL -> "лунном камне";
        };
    }

    @Override
    public String toString() {
        return "лунный камень";
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (!(o instanceof MoonStone moonStone)) return false;
        if (!super.equals(o)) return false;
        return Objects.equals(moonStone.toString(), moonStone.toString());
    }

    @Override
    public int hashCode() {
        return Objects.hash(super.hashCode(), toString());
    }
}

File: proga/1_semestr/labs/lab3/src/main/kotlin/dev/lysmux/lab3/entity/item/MoonStone.kt
package dev.lysmux.lab3.entity.item

import dev.lysmux.lab3.common.Case

class MoonStone : Item(){
    override fun caseDeclension(toCase: Case): String {
        return when (toCase) {
            Case.NOMINATIVE, Case.ACCUSATIVE -> "лунный камень"
            Case.GENITIVE -> "лунного камня"
            Case.DATIVE -> "лунному камню"
            Case.CREATIVE -> "лунным камнем"
            Case.PREPOSITIONAL -> "лунном камне"
        }
    }

    override fun getPluralName() = "лунные камни"

    override fun toString() = "лунный камень"

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (javaClass != other?.javaClass) return false
        return true
    }

    override fun hashCode(): Int {
        return javaClass.hashCode()
    }
}